pkg/controllers/metrics/node/counter/counter.go

```diff
@@ -0,0 +1,166 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package counter
+
+import (
+	"context"
+	"reflect"
+	"strings"
+
+	karpenterapi "github.com/awslabs/karpenter/pkg/apis/provisioning/v1alpha4"
+	"github.com/awslabs/karpenter/pkg/metrics"
+	"github.com/prometheus/client_golang/prometheus"
+	coreapi "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"knative.dev/pkg/logging"
+	crmetrics "sigs.k8s.io/controller-runtime/pkg/metrics"
+)
+
+const (
+	metricArchLabel                    = "arch"
+	metricConditionDiskPressureLabel   = "diskpressure"
+	metricConditionMemoryPressureLabel = "memorypressure"
+	metricConditionPIDPressureLabel    = "pidpressure"
+	metricConditionReadyLabel          = "ready"
+	metricInstanceTypeLabel            = "instancetype"
+	metricTopologyRegionLabel          = "region"
+	metricTopologyZoneLabel            = "zone"
+	metricOsLabel                      = "os"
+)
+
+var (
+	nodeCountGaugeVec = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: metrics.KarpenterNamespace,
+			Subsystem: "cluster",
+			Name:      "node_count",
+			Help:      "Count of cluster nodes. Broken out by topology and status.",
+		},
+		[]string{
+			metrics.ProvisionerLabel,
+			metricArchLabel,
+			metricConditionDiskPressureLabel,
+			metricConditionMemoryPressureLabel,
+			metricConditionPIDPressureLabel,
+			metricConditionReadyLabel,
+			metricInstanceTypeLabel,
+			metricTopologyRegionLabel,
+			metricTopologyZoneLabel,
+			metricOsLabel,
+		},
+	)
+
+	prometheusLabelsFor = make(map[types.NamespacedName]prometheus.Labels)
+
+	conditionTypeToMetricLabel = map[coreapi.NodeConditionType]string{
+		coreapi.NodeDiskPressure:   metricConditionDiskPressureLabel,
+		coreapi.NodeMemoryPressure: metricConditionMemoryPressureLabel,
+		coreapi.NodePIDPressure:    metricConditionPIDPressureLabel,
+		coreapi.NodeReady:          metricConditionReadyLabel,
+	}
+)
+
+func init() {
+	crmetrics.Registry.MustRegister(nodeCountGaugeVec)
+}
+
+// UpdateCount updates the emitted metric based on the node's current status relative to the
+// past status. If the data for `node` cannot be populated then `nil` should be passed as the
+// argument.
+func UpdateCount(ctx context.Context, name types.NamespacedName, node *coreapi.Node) {
```

Comment:
Consider counter.Update() instead of counter.UpdateCount(), which stutters. You can also get the entire namespaced name from the node object `node.Name` (every kubernetes object has it). In this case, nodes are a cluster global resource -- namespace for nodes will never make sense.